evap/grades/views.py

```diff
@@ -0,0 +1,174 @@
+from django.utils.translation import ugettext as _
+from django.http import HttpResponseBadRequest, HttpResponseForbidden
+from django.contrib.auth.decorators import login_required
+
+import os
+from sendfile import sendfile
+
+from evap.evaluation.auth import grade_publisher_required, grade_downloader_required
+from evap.evaluation.models import Semester, Contribution, Course
+from evap.grades.models import GradeDocument
+from evap.grades.forms import GradeDocumentForm
+from evap.evaluation.tools import send_publish_notifications
+
+def get_graded_courses_with_prefetched_data(semester):
+    courses = semester.course_set.filter(is_graded=True).exclude(state='new').prefetch_related(
+        Prefetch("contributions", queryset=Contribution.objects.filter(responsible=True).select_related("contributor"), to_attr="responsible_contribution"),
+        "degrees")
+
+    course_data = []
+    for course in courses:
+        course.responsible_contributor = course.responsible_contribution[0].contributor
+        course_data.append((course, GradeDocument.objects.filter(course=course, type=GradeDocument.FINAL_GRADES).exists()))
+
+    return course_data
+
+
+@grade_publisher_required
+def index(request):
+    template_data = dict(
+        semesters=Semester.objects.all()
+    )
+    return render(request, "grades_index.html", template_data)
+
+
+@grade_publisher_required
+def semester_view(request, semester_id):
+    semester = get_object_or_404(Semester, id=semester_id)
+
+    courses = get_graded_courses_with_prefetched_data(semester)
+
+    template_data = dict(
+        semester=semester,
+        courses=courses,
+        disable_if_archived="disabled=disabled" if semester.is_archived else "",
+        disable_breadcrumb_semester=True,
+    )
+    return render(request, "grades_semester_view.html", template_data)
+
+
+@grade_publisher_required
+def course_view(request, semester_id, course_id):
+    semester = get_object_or_404(Semester, id=semester_id)
+    course = get_object_or_404(Course, id=course_id)
+
+    grade_documents = GradeDocument.objects.filter(course=course)
+
+    template_data = dict(
+        semester=semester,
+        course=course,
+        grade_documents=grade_documents,
+        disable_if_archived="disabled=disabled" if semester.is_archived else "",
+        disable_breadcrumb_course=True,
+    )
+    return render(request, "grades_course_view.html", template_data)
+
+
+def helper_grade_upload(request, course, final_grades=False, instance=None):
+    created = True
+    if request.method == "POST":
+        if instance:
+            created = False
+        form = GradeDocumentForm(request.POST, request.FILES, instance=instance)
+        if form.is_valid():
+            instance = form.save(commit=False)
+            instance.last_modified_user = request.user
+            instance.course = course
+            if final_grades:
+                instance.type = GradeDocument.FINAL_GRADES
+            instance.save()
```

Comment:
the instance type must be set in advance, otherwise the form wouldn't be valid at all. the user can't select the type by hand so it is set in this method.